Sources/_SwiftSyntaxMacros/ExpressionMacro.swift

```diff
@@ -20,3 +24,54 @@ public protocol ExpressionMacro: Macro {
     _ macro: MacroExpansionExprSyntax, in context: MacroEvaluationContext
   ) -> MacroResult<ExprSyntax>
 }
+
+#if canImport(_CompilerPluginSupport)
+extension ExpressionMacro {
+  public static func _kind() -> _CompilerPluginKind {
+    .expressionMacro
+  }
+
+  public static func _rewrite(
+    targetModuleName: UnsafePointer<UInt8>,
+    targetModuleNameCount: Int,
+    filePath: UnsafePointer<UInt8>,
+    filePathCount: Int,
+    sourceFileText: UnsafePointer<UInt8>,
+    sourceFileTextCount: Int,
+    localSourceText: UnsafePointer<UInt8>,
+    localSourceTextCount: Int
+  ) -> (UnsafePointer<UInt8>?, count: Int) {
+    let targetModuleNameBuffer = UnsafeBufferPointer(
+      start: filePath, count: targetModuleNameCount)
+    let targetModuleName = String(
+      decoding: targetModuleNameBuffer, as: UTF8.self)
+    let filePathBuffer = UnsafeBufferPointer(
+      start: filePath, count: filePathCount)
+    let filePath = String(decoding: filePathBuffer, as: UTF8.self)
+    let sourceFileTextBuffer = UnsafeBufferPointer(
+      start: sourceFileText, count: sourceFileTextCount)
+    let sourceFileText = String(decoding: sourceFileTextBuffer, as: UTF8.self)
+    // TODO: `SourceLocationConverter.init(file:source:)` should be able to
+    // accept a buffer.
+    let converter = SourceLocationConverter(
+      file: filePath, source: sourceFileText)
+    let context = MacroEvaluationContext(
+      moduleName: targetModuleName, sourceLocationConverter: converter)
+    let meeTextBuffer = UnsafeBufferPointer(
+      start: localSourceText, count: localSourceTextCount)
+    var parser = Parser(meeTextBuffer)
+    let mee = try! MacroExpansionExprSyntax.parse(from: &parser)
```

Comment:
To get correct source-location and enclosing-context information, one really needs to parse the whole source file and then dig out the expansion node that's going to be expanded. There's a utility over in ASTGen that can do this based on the UTF-8 offset in the buffer, which I was planning to bring over to swift-syntax because it's generally useful. It's at https://github.com/apple/swift/blob/main/lib/ASTGen/Sources/ASTGen/Macros.swift#L5-L20

Reply:
I went ahead and landed this as https://github.com/apple/swift-syntax/pull/1031

Reply:
Thanks, fixed.